src/core/ComboBox/ComboBox.tsx

```diff
@@ -0,0 +1,318 @@
+    inputValue: string,
+  ) => SelectOption<T>[];
+  /**
+   * Native input element props.
+   */
+  inputProps?: Omit<React.ComponentPropsWithoutRef<'input'>, 'size'>;
+  /**
+   * Props to customize dropdown menu behavior.
+   */
+  dropdownMenuProps?: PopoverProps;
+} & Omit<CommonProps, 'title'>;
+
+/**
+ * ComboBox component that allows typing a value to filter the options in dropdown list.
+ * Values can be selected either using mouse clicks or using the Enter key.
+ * @example
+ * <ComboBox
+ *   options={[
+ *     { label: 'Item 1', value: 1 },
+ *     { label: 'Item 2', value: 2 },
+ *     { label: 'Item 3', value: 3 },
+ *   ]}
+ *   onChange={() => {}}
+ * />
+ */
+export const ComboBox = <T,>(props: ComboBoxProps<T>) => {
+  const {
+    options,
+    value,
+    onChange,
+    filterFunction,
+    className,
+    inputProps,
+    dropdownMenuProps,
+    ...rest
+  } = props;
+
+  // Generate a stateful random id if not specified
+  const [id] = React.useState(() => props.id ?? `iui-${nanoid(10)}`);
+
+  useTheme();
+
+  /** Generates an memoized id for an option, given the index from original list */
+  const getOptionId = React.useCallback(
+    (index: number) =>
+      options[index].id ??
+      `${id}-option${options.findIndex(
+        ({ value }) => value === options[index].value,
+      )}`,
+    [options, id],
+  );
+
+  const inputRef = React.useRef<HTMLInputElement>(null);
+  const menuRef = React.useRef<HTMLUListElement>(null);
+  const [isOpen, setIsOpen] = React.useState(false);
+
+  // Set min-width of menu to be same as input
+  const [minWidth, setMinWidth] = React.useState(0);
+  React.useEffect(() => {
+    if (inputRef.current) {
+      setMinWidth(inputRef.current.offsetWidth);
+    }
+  }, [isOpen]);
+
+  const [filteredOptions, setFilteredOptions] = React.useState(options);
+  const [focusedIndex, setFocusedIndex] = React.useState<number>(-1);
+
+  // Maintain internal selected value state synced with `value` prop
+  const [selectedValue, setSelectedValue] = React.useState<T | undefined>();
+  React.useEffect(() => {
+    setSelectedValue(value);
+  }, [value]);
+
+  // Controlled input value
+  const [inputValue, setInputValue] = React.useState<string>('');
+  const onInput = React.useCallback(
+    (event: React.ChangeEvent<HTMLInputElement>) => {
+      setInputValue(event.target.value);
+      inputProps?.onChange?.(event);
+    },
+    [inputProps],
+  );
+
+  // Fire onChange callback and update inputValue every time selected value changes
+  React.useEffect(() => {
+    if (selectedValue) {
+      onChange?.(selectedValue);
+      setInputValue(
+        options.find(({ value }) => value === selectedValue)?.label ?? '',
+      );
+    }
+  }, [selectedValue, onChange, options]);
+
+  // Filter options and update focus when input value changes
+  React.useEffect(() => {
+    const selectedOption = options.find(({ value }) => value === selectedValue);
+    if (
+      !inputValue ||
+      !isOpen ||
```

Comment:
To optimize even more, you can check if it is open and if not then you don't need to do anything, including searching for `selectedOption`

Reply:
Done